src/ansi-c/c_typecheck_code.cpp

```diff
@@ -995,34 +995,34 @@ void c_typecheck_baset::typecheck_spec_assigns_target(exprt &target)
   else if(can_cast_expr<side_effect_expr_function_callt>(target))
   {
     const auto &funcall = to_side_effect_expr_function_call(target);
-    if(can_cast_expr<symbol_exprt>(funcall.function()))
+    if(!can_cast_expr<symbol_exprt>(funcall.function()))
     {
-      const auto &ident = to_symbol_expr(funcall.function()).get_identifier();
-      if(
-        ident == CPROVER_PREFIX "assignable" ||
-        ident == CPROVER_PREFIX "object_whole" ||
-        ident == CPROVER_PREFIX "object_upto" ||
-        ident == CPROVER_PREFIX "object_from")
-      {
-        for(const auto &argument : funcall.arguments())
-          throw_on_side_effects(argument);
-        return;
-      }
+      throw invalid_source_file_exceptiont(
+        "function pointer calls not allowed in assigns clauses",
+        target.source_location());
     }
-    else
+
+    if(target.type().id() != ID_empty)
     {
       throw invalid_source_file_exceptiont(
-        "function pointer calls not allowed in assigns clauses",
+        "expecting void return type for function '" +
+          id2string(to_symbol_expr(funcall.function()).get_identifier()) +
+          "' called in assigns clause",
         target.source_location());
     }
+
+    for(const auto &argument : funcall.arguments())
+      throw_on_side_effects(argument);
+  }
+  else
+  {
+    // if we reach this point the target did not pass the checks
+    throw invalid_source_file_exceptiont(
+      "assigns clause target must be a non-void lvalue, a call "
+      "to " CPROVER_PREFIX
+      "POINTER_OBJECT or a call to a function returning void",
```

Comment:
Why do we need to remove here all `CPROVER_PREFIX_object_*` cases? Make the error message shorter? For me, it was quite informative before.